sdk/src/Services/SQS/Generated/Model/Internal/MarshallTransformations/AddPermissionRequestMarshaller.cs

```diff
@@ -63,54 +63,58 @@ public IRequest Marshall(AddPermissionRequest publicRequest)
             request.HttpMethod = "POST";
 
             request.ResourcePath = "/";
-            using (MemoryStream memoryStream = new MemoryStream())
+#if NETCOREAPP3_1_OR_GREATER
```

Comment:
This might be stretching the scope of this change too much, but would it be possible to make the source of `MemoryStream`/`IBufferWriter<T>` pluggable.

I'd really love to add my own pipeline middleware that allows for injecting a [RecyclableMemoryStream](https://github.com/microsoft/Microsoft.IO.RecyclableMemoryStream) (which implements both `MemoryStream` and `IBufferWriter<T>`)

Reply:
My interpretation of this is we would have a property on `AWSConfigs` called `StreamFactory` of type `IStreamFactory` defined as:
```
public interface IStreamFactory
{
    Stream GetStream();
}
```

For .NET Core 3.1 and above we would default this to an instance of ArrayBufferWriterFactory:
```
internal class ArrayBufferWriterFactory : IStreamFactory
{
    public Stream GetStream()
    {
        return ArrayBufferWriter<byte>();
    }
}
```

For older targets we would use a MemoryStream based factory:
```
internal class MemoryStreamFactory : IStreamFactory
{
    public Stream GetStream()
    {
        throw new MemoryStream();
    }
}
```

In our unmarshallers where we need a stream we could get rid of the conditional check and just call `AWSConfigs.StreamFactory.GetManager()`.

Then @slang25 can take a dependency on `Microsoft.IO.RecyclableMemoryStream` and right his own factory and assign it to `AWSConfigs.StreamFactory`.
```
internal class RecyclableMemoryStreamFactory : IStreamFactory
{
    private static readonly RecyclableMemoryStreamManager manager = new RecyclableMemoryStreamManager();

    public Stream GetStream()
    {
        return manager.GetStream();
    }
}
```

Does that capture your thoughts @slang25 ?

Reply:
Interesting, I just took a look at the class linked that does something similar, which allows the a wrapper over IBufferWriter and at first glance it doesn't seem to bad, but i'm not sure why they are overriding `Write` and `ReadAsync` etc.. IBufferWriter only defines 3 methods to implement.

IF we don't have to override all those methods listed [here](https://github.com/CommunityToolkit/dotnet/blob/c23e1cc6e918f86c589facf98f70ca122495c385/src/CommunityToolkit.HighPerformance/Streams/IBufferWriterStream%7BTWriter%7D.cs) I think it is doable. I'll hold off until I hear back though.

Reply:
What I'm looking for is a way to get the best perf out of the AWS SDK, so having the ability to pool these buffers that are going to copy all the bytes into memory is going to reduce memory allocations significantly.

My understanding is that `IBufferWriter<T>` is the fast modern to do [synchronous buffered writing](https://learn.microsoft.com/en-us/dotnet/standard/io/buffers#ibufferwritert), vs `Stream` which is more ubiquitous in the .NET world today.

That said, I think a `Stream GetStream()` signature would work, but you could additional do a type check to see if it implements a `IBufferWriter<T>`, and then use the appropriate overloads. That makes it convenient for use with Streams, and has a fast path for things like `RecyclableMemoryStreamManager`. To be honest, this isn't a deal breaker, I haven't ran benchmarks but I'm guessing the gains are not tremendous.

The problem with a static `IStreamFactory` is that I want to be able to pool the buffers and therefore want a way to return them. So the `AWSConfigs` static property would need an interface to support that for it to achieve what I'm looking for.

I was thinking about having a `PipelineHandler` that could be added into the pipeline, and setting the stream on some sort of context object that passes through. That way it could own the creation/pooling of buffers, and returning them. However a static property would be more ergonomic.